### ISSUE TITLE
Segfault when moving file when using --monitor

### ISSUE THREAD

**Issue by MichaelGissingNC** [abraunegg/onedrive#995]
# Environment

I tried two different images:

* self built
  * ``docker run --name "onedrive" -it -v `pwd`/conf:/onedrive/conf -v `pwd`/data:/onedrive/data -e "ONEDRIVE_UID:`id -u`" -e "ONEDRIVE_GID:`id -g`" local-onedrive``
* `driveone/onedrive:latest`
  * ``docker run --name "onedrive" -it -v `pwd`/conf:/onedrive/conf -v `pwd`/data:/onedrive/data -e "ONEDRIVE_UID:`id -u`" -e "ONEDRIVE_GID:`id -g`" driveone/onedrive``

# Behavior

* In both cases, log-in and initial sync runs OK.
* Creating a new files is OK, it gets uploaded to OneDrive
* Moving a file crashes the container (`mv abc xyz`)
  * process terminates, no log message
  * In host journal I get the following message:
    * for local compiled: `kernel: onedrive[11281]: segfault at 48 ip 00000000006f3c64 sp 00007fff63e310f8 error 4 in onedrive[400000+3a8000]`
    * for driveone/onedrive:latest: `kernel: onedrive[12750]: segfault at 48 ip 00000000006f3c64 sp 00007ffeeca1f748 error 4 in onedrive[400000+3a8000]`

**Comment by abraunegg** [comment]
@MichaelGissingNC 
Please can you follow the correct support procedures as outlined here: https://github.com/abraunegg/onedrive#reporting-issues

You have not used the bug report template to submit the above issue - this is why the bug report template exists - to provide support in a timely fashion.

Please update your post with all the required details, include a CORRECT debug log that SHOWS the error and segfault occurring and provide what Docker version you are running.

Additionally, please perform your tests without using Docker, as this will assist in determining if this is a Docker related issue or bug in the code.

**Comment by MichaelGissingNC** [comment]
@abraunegg thanks for the quick response

First, let me respond to your direct comments

* I can't provide you with the full debug log output as I can't determine which encoded data carries confidential information and also the included paths and file names themselves are confidential
* I can provide any requested parts of the debug log if you tell me what you are looking for after removing any sensitive information
* You don't see the segfault in the docker / onedrive logs, I failed to get a more verbose description of the segfault somewhere. The only indication I see is in the host's journal.
* I'll try to run the bare client without docker in a VM later, I won't (for various reasons) run it directly.

Now I try to provide all the information in a template compatible way as well:

### Bug Report Details ###
**Describe the bug**
Moving a file causes onedrive (docker) to crash.

**Application and Operating System Details:**
*   OS: `Linux HOST 4.15.0-20-generic #21-Ubuntu SMP Tue Apr 24 06:16:15 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`, `Linux Mint 19 Tara`
*   `GUI present`
*   OneDrive Account Type: `Business`
*   Did you build from source or install from a package? `Docker driveone/onedrive:latest`
*   Application configuration: Output of `onedrive --display-config`
    ```
    onedrive version                       = v2.4.3-11-g2b7e36d
    Config path                            = /root/.config/onedrive
    Config file found in config path       = false
    Config option 'check_nosync'           = false
    Config option 'sync_dir'               = /root/OneDrive
    Config option 'skip_dir'               = 
    Config option 'skip_file'              = ~*|.~*|*.tmp
    Config option 'skip_dotfiles'          = false
    Config option 'skip_symlinks'          = false
    Config option 'monitor_interval'       = 300
    Config option 'min_notify_changes'     = 5
    Config option 'log_dir'                = /var/log/onedrive/
    Config option 'classify_as_big_delete' = 1000
    Config option 'sync_root_files'        = false
    Selective sync 'sync_list' configured  = false
    Business Shared Folders configured     = false
    ```
*   Curl Version: Output of `curl --version`
    ```
    curl 7.29.0 (x86_64-redhat-linux-gnu) libcurl/7.29.0 NSS/3.44 zlib/1.2.7 libidn/1.28 libssh2/1.8.0
    Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtsp scp sftp smtp smtps telnet tftp 
    Features: AsynchDNS GSS-Negotiate IDN IPv6 Largefile NTLM NTLM_WB SSL libz unix-sockets
    ```

**To Reproduce**
Steps to reproduce the behavior if not causing an application crash:
1. create a new directory (`~/OneDrive3` in this case), with children `conf`, `data` and `log`
1. run ``docker run --name "onedrive3" -it -v `pwd`/conf:/onedrive/conf -v `pwd`/data:/onedrive/data -v `pwd`/log:/var/log/onedrive -e "ONEDRIVE_UID:`id -u`" -e "ONEDRIVE_GID:`id -g`" -e ONEDRIVE_DEBUG=1 driveone/onedrive``
2. Do the authorization (copy from browser)
3. Initial sync starts
4. Initial sync finishes
5. wait some time (not necessary)
6. create a file
7. move the created file
8. crash

**Complete Verbose Log Output**
Not sure how to provide this, given it's in docker.

Application Log Output:
See Screenshot

**Screenshots**

![Selection_00648](https://user-images.githubusercontent.com/27948276/87927028-d701cc80-ca82-11ea-8e4c-e64a04075b2e.png)

1. Initial sync has completed
2. I waited a few moments (1-2 minutes)
3. I created a file called `abc` in the directory via `touch abc`
4. I waited another bit after step 3 completed and then moved the file to be named `abcde` via `mv abc abcde`
5. No additional log output, container terminates here
6. A message indicating a segfault appears in host's journal (not on screenshot)

### Bug Report Checklist ###
*   [+] Detailed description
*   [+] Reproduction steps (if applicable)
*   [-] Verbose Log Output

**Comment by abraunegg** [comment]
@MichaelGissingNC 
I am able to reproduce this locally, without Docker, when using a default install of Linux Mint 19 MATE.

**Comment by abraunegg** [comment]
@MichaelGissingNC 

Please can you help validate the following PR to resolve this issue:
```
git clone https://github.com/abraunegg/onedrive.git
cd onedrive
git fetch origin pull/997/head:pr997
git checkout pr997
./configure; make clean; make;
```

You will need to manually build an updated Docker container from this PR to test under Docker, however in my testing the PR resolves the issue you have encountered.

**Comment by MichaelGissingNC** [comment]
@abraunegg thank you for the quick fix. It works for me.

1. checked out latest commit (`40ee78ee29`)
1. built new image
1. created new directory to sync to
1. initial sync
1. - [x] create file and move it
1. - [x] file is correctly available in OneDrive web interface
1. - [x] onedrive container still running, continuing to monitor

**Comment by abraunegg** [comment]
@MichaelGissingNC 
Please can you let me know how you go with your testing - please test all files / folder combinations. I did that this morning, ran it through also my test scripts and no issue locally. 

The thing is - the commit that caused this, I also tested .. so something went west 14 days ago with that particular commit & it's testing that I failed to pick up - lack of proper testing is to blame for this one (#711) - I am sorry for the inconvenience that this bug has caused you.

**Comment by abraunegg** [comment]
@MichaelGissingNC 
PR merged into master, closing issue

**Comment by norbusan** [comment]
@abraunegg bit late to the game, but checked now through the changes. Does that mean that `isFile` crashed? That would be surprising, or? But at least this is what I was reading from your changes.

**Comment by abraunegg** [comment]
Correct .. isDir and isFile was now causing a segfault .. it should not have caused a segfault, but for whatever reason it was

**Comment by norbusan** [comment]
This is **wicked** and a bug in the D libraries I guess...

**Comment by MichaelGissingNC** [comment]
@abraunegg I can't really provide structured testing, but I have the latest version now in use on my machine and I'll keep an eye on it.
